src/meta/src/rpc/server.rs

```diff
@@ -228,7 +243,13 @@ pub async fn rpc_serve_with_store<S: MetaStore>(
         .expect("Unable to start leader services");
     });
 
-    Ok((join_handle, svc_shutdown_tx))
+    // kill everything on panic signal
+    let join_handle_with_trigger = tokio::spawn(async move {
+        let _ = panic_rx.changed().await;
+        join_handle.abort();
```

Comment:
The timing of the abort is still uncertain, I tried to panic in `run_elections` and use `#[should_panic(expected = "This node lost its leadership. Exiting node")]` for `test_fencing`. It didn't work, because the handle in `run_elections` doesn't join back, its panic didn't effect the thread of leader RPC service. So there will be two or more leader RPC services exist during test:
```
Expected to have 1 leader, instead got 2 leaders
Left:  2
Right: 1

thread 'rpc::server::tests::test_fencing' panicked at 'This node lost its leadership. Exiting node', src/meta/src/rpc/elections.rs:400:33
stack backtrace:
   0: rust_begin_unwind
             at /rustc/bdb07a8ec8e77aa10fb84fae1d4ff71c21180bb4/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/bdb07a8ec8e77aa10fb84fae1d4ff71c21180bb4/library/core/src/panicking.rs:64:14
   2: risingwave_meta::rpc::elections::run_elections::{{closure}}::{{closure}}
             at ./src/rpc/elections.rs:400:33
   3: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.37/src/instrument.rs:272:9
   4: tokio::runtime::task::core::Core<T,S>::poll::{{closure}}
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/task/core.rs:208:17
   5: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/loom/std/unsafe_cell.rs:14:9
   6: tokio::runtime::task::core::Core<T,S>::poll
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/task/core.rs:197:13
   7: tokio::runtime::task::harness::poll_future::{{closure}}
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/task/harness.rs:483:19
   8: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/bdb07a8ec8e77aa10fb84fae1d4ff71c21180bb4/library/core/src/panic/unwind_safe.rs:271:9
   9: std::panicking::try::do_call
             at /rustc/bdb07a8ec8e77aa10fb84fae1d4ff71c21180bb4/library/std/src/panicking.rs:483:40
  10: ___rust_try
  11: std::panicking::try
             at /rustc/bdb07a8ec8e77aa10fb84fae1d4ff71c21180bb4/library/std/src/panicking.rs:447:19
  12: std::panic::catch_unwind
             at /rustc/bdb07a8ec8e77aa10fb84fae1d4ff71c21180bb4/library/std/src/panic.rs:137:14
  13: tokio::runtime::task::harness::poll_future
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/task/harness.rs:471:18
  14: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/task/harness.rs:107:27
  15: tokio::runtime::task::harness::Harness<T,S>::poll
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/task/harness.rs:61:15
  16: tokio::runtime::task::raw::poll
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/task/raw.rs:194:5
  17: tokio::runtime::task::raw::RawTask::poll
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/task/raw.rs:134:18
  18: tokio::runtime::task::LocalNotified<S>::run
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/task/mod.rs:430:9
  19: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}::{{closure}}
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/scheduler/current_thread.rs:570:25
  20: tokio::runtime::coop::with_budget
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/coop.rs:102:5
  21: tokio::runtime::coop::budget
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/coop.rs:68:5
  22: tokio::runtime::scheduler::current_thread::Context::run_task::{{closure}}
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/scheduler/current_thread.rs:277:29
  23: tokio::runtime::scheduler::current_thread::Context::enter
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/scheduler/current_thread.rs:340:19
  24: tokio::runtime::scheduler::current_thread::Context::run_task
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/scheduler/current_thread.rs:277:9
  25: tokio::runtime::scheduler::current_thread::CoreGuard::block_on::{{closure}}
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/scheduler/current_thread.rs:569:34
  26: tokio::runtime::scheduler::current_thread::CoreGuard::enter::{{closure}}
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/scheduler/current_thread.rs:601:57
  27: tokio::macros::scoped_tls::ScopedKey<T>::set
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/macros/scoped_tls.rs:61:9
  28: tokio::runtime::scheduler::current_thread::CoreGuard::enter
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/scheduler/current_thread.rs:601:27
  29: tokio::runtime::scheduler::current_thread::CoreGuard::block_on
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/scheduler/current_thread.rs:520:19
  30: tokio::runtime::scheduler::current_thread::CurrentThread::block_on
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/scheduler/current_thread.rs:154:24
  31: tokio::runtime::runtime::Runtime::block_on
             at /Users/august/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.22.0/src/runtime/runtime.rs:279:47
  32: risingwave_meta::rpc::server::tests::test_fencing
             at ./src/rpc/server.rs:498:9
  33: risingwave_meta::rpc::server::tests::test_fencing::{{closure}}
             at ./src/rpc/server.rs:454:29
  34: core::ops::function::FnOnce::call_once
             at /rustc/bdb07a8ec8e77aa10fb84fae1d4ff71c21180bb4/library/core/src/ops/function.rs:507:5
  35: risingwave_test_runner::test_runner::run_test_inner::{{closure}}::{{closure}}
             at /Users/august/Documents/code/risingwave/src/test_runner/src/test_runner.rs:67:21
  36: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/bdb07a8ec8e77aa10fb84fae1d4ff71c21180bb4/library/core/src/ops/function.rs:507:5
  37: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/bdb07a8ec8e77aa10fb84fae1d4ff71c21180bb4/library/alloc/src/boxed.rs:2000:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
leader: http://127.0.0.1:1600, resp: Response { metadata: MetadataMap { headers: {"content-type": "application/grpc", "date": "Fri, 06 Jan 2023 09:12:21 GMT", "grpc-status": "0"} }, message: AddWorkerNodeResponse { status: None, node: Some(WorkerNode { id: 1, r#type: ComputeNode, host: Some(HostAddress { host: "127.0.0.1", port: 1700 }), state: Starting, parallel_units: [ParallelUnit { id: 0, worker_node_id: 1 }, ParallelUnit { id: 1, worker_node_id: 1 }, ParallelUnit { id: 2, worker_node_id: 1 }, ParallelUnit { id: 3, worker_node_id: 1 }, ParallelUnit { id: 4, worker_node_id: 1 }] }) }, extensions: Extensions }
leader: http://127.0.0.1:1602, resp: Response { metadata: MetadataMap { headers: {"content-type": "application/grpc", "date": "Fri, 06 Jan 2023 09:12:21 GMT", "grpc-status": "0"} }, message: AddWorkerNodeResponse { status: None, node: Some(WorkerNode { id: 1001, r#type: ComputeNode, host: Some(HostAddress { host: "127.0.0.1", port: 1702 }), state: Starting, parallel_units: [ParallelUnit { id: 1000, worker_node_id: 1001 }, ParallelUnit { id: 1001, worker_node_id: 1001 }, ParallelUnit { id: 1002, worker_node_id: 1001 }, ParallelUnit { id: 1003, worker_node_id: 1001 }, ParallelUnit { id: 1004, worker_node_id: 1001 }] }) }, extensions: Extensions }
thread 'rpc::server::tests::test_fencing' panicked at 'assertion failed: `(left == right)`
  left: `2`,
 right: `1`: Expected to have 1 leader, instead got 2 leaders', src/meta/src/rpc/server.rs:491:13
stack backtrace:
```

Reply:
Let's hold this PR and the other one, @shanicky has almost finished the election refactoring. Let's see if it can be well handled there.

Reply:
I agree, we still have some uncertainty, since we periodically check against the meta store what the status of the leader is. I never ran into the above issue. Did the current implementation lead to the test failure or did the test fail when you used panic?

Reply:
Used panic.

Reply:
True. If you use panic in the election part it will not kill the node. However the abort signal should kill the services, should it not? IMHO the current implementation is valid. Even if we use the election refactoring by @shanicky we will still need a fencing mechanism, right?